src/volume/volume.cpp

```diff
@@ -162,44 +165,58 @@ float Volume::weight(float x)
 // This functions returns the results of a cubic interpolation using 4 values and a factor
 float Volume::cubicInterpolate(float g0, float g1, float g2, float g3, float factor)
 {
-    return g0 * pow(factor, 3) + g1 * pow(factor, 2) + g2 * factor + g3;
+    float value = pow(factor, 3) * (-0.5 * g0 + 1.5 * g1 - 1.5 * g2 + 0.5 * g3) + pow(factor, 2) * (g0 - 2.5 * g1 + 2 * g2 - 0.5 * g3) + factor * (-0.5 * g0 + 0.5 * g2) + g1;
+    return value;
 }
 
 // ======= TODO : Check if correct ========
 // This function returns the value of a bicubic interpolation
 float Volume::bicubicInterpolateXY(const glm::vec2& xyCoord, int z) const
 {
-    glm::mat4 values;
-    glm::vec4 dist_x;
-    glm::vec4 dist_y;
+    glm::vec4 values;
 
     int offset = 1;
-    for (int i = 0 - offset; i < 4 - offset; i++) {
-        for (int j = 0 - offset; j < 4 - offset; j++) {
-            int x = static_cast<int>(xyCoord.x + i);
-            int y = static_cast<int>(xyCoord.y + j);
+    // int y = static_cast<int>(xyCoord.y);
 
-            values[i + offset][j + offset] = this->getVoxel(x, y, z);
-            dist_x[i + offset] = this->weight(xyCoord.x - x);
-            dist_y[j + offset] = this->weight(xyCoord.y - y);
-        }
-    }
+    const int x = static_cast<int>(xyCoord.x);
+    const int y = static_cast<int>(xyCoord.y);
+
+    const float fac_x = xyCoord.x - float(x);
+    const float fac_y = xyCoord.y - float(y);
 
-    glm::vec1 result = (dist_x * values) * dist_y;
+    for (int i = 0 - offset; i < 4 - offset; i++) {

```

Comment:
edge conditions can be added here, as code gives error when the image is adjusted.

Reply:
The values are checked [here](https://github.com/MihaiAnton/volume-visualization/blob/3d93d4cad62d0bb6505b3445aba8f0e1901354e5/src/volume/volume.cpp#L81). Is this the possible error you are referring to?